### DISCUSSION TITLE
Passing Updated Variables into Subsequent Animation Steps

### DISCUSSION THREAD

**Discussion by dusevitch** [sofa-framework/sofa#4125] [Write / Run a simulation] [answered]
Can't quite figure out how to pass variables into subsequent animations. In the below code I'm trying to run `animation_up` after `animation_left` but  the animation jumps back to the position it started at to run the second animation (aka the position of actuators go back to its original position). 

This could be a python scoping issue, but doesn't seem like it. Even testing just one `animate` function  results in updated variable `actuators` with the proper value, so I'm guessing its some issue with the timing of when these functions are called?

How do I properly pass an updated `actuators` value in? None of my attempts have worked (a few examples of attempts in the code snippet below).

```python
def animation_left( actuators, step, factor):
    rigid1 = RigidDof(actuators[1].ServoMotor.ServoBody.dofs)
    rigid1.setPosition(rigid1.rest_position + rigid1.left * step * factor)

def animation_up( actuators, step, factor):   
    rigid = RigidDof(actuators[1].ServoMotor.ServoBody.dofs)
    rigid.setPosition(rigid.rest_position - rigid.forward * step * factor) 

def runAnimation(actuators, step): 
    def startAnimation2(actuator_val, step, factor):
          animate(animation_up,{"actuators": actuator_val, "step": step}, duration=0.2)      
      
    # multiple attempts (I've run the below lines separately with minor other changes to the code): 
    animate(animation_left,{"actuators": actuators, "step": step}, duration=0.2, onDone=startAnimation2()
    animate(animation_left,{"actuators": actuators, "step": step}, duration=0.2, onDone=startAnimation2(factor=1.0, **dict(actuator_val=actuators, step=step)))


def initFunc(self, key):
    if key == Key.X:
        runAnimation(self.actuators, 70.0)      
```

Checking `animate.py` in splib3 the doOnDone seems like it's expecting a function call taking in parameters, hence my above attempt to input them.

```python
def doOnDone(self, currentTime):
    self.onDone(factor=self.factor, **self.params)
```

Running animations separately of course results in non sequential behavior which was solved by @damienmarchal [here](https://github.com/sofa-framework/sofa/discussions/4101).

```python
animate(animation_left,{"actuators": actuators, "step": step}, duration=0.2)
animate(animation_up,{"actuators": actuators, "step": step}, duration=0.2)
```

I think following a path for an animation is something that could be of high interest to the community but seems to me like it can't really be performed currently without complicated iterations. I may be able to contribute something to do this but I need to better understand the basics of animations.

---
** **My application for reference:** ** Trying to run an animation prior to commanding actuators to control a material. I'm trying to animate a 'preloading' of a springy material before controlling it via actuators. This is essentially because I don't know the exact forces applied to the model and I'm trying to test multiple desigs and iterations (eventually run an optimization). Any other suggestions to do this are welcome.

**Comment by damienmarchal** [comment, marked as answer]
@dusevitch,

The problem is because in the two animation you are computing the new position from the rest_position. In Sofa the rest_position is the position of the object when the scene was loaded so. So the behavior you are noticing is the one that is described in the code sample you provide. 

If you want to use a new reference position in animation_up, the simplest is probably to add it as a parameter to the function. 
Here is a self-contained example that fix all that things and show how to chain two function while sharing state using an extra parameter in the animate_up section. 

```python
import Sofa
import splib3 
from splib3.animation.animate import AnimationManager, animate
from splib3.numerics import RigidDof

def animation_left(target, step, factor):
    rigid = RigidDof(target.dofs)
    rigid.setPosition(rigid.rest_position + rigid.left * step * factor)
    print(f"Left at ({factor:0.1}), position is: {rigid.getPosition()}")

def animation_up(target, start_position, step, factor):   
    """the start position.... is the base location where the computation will happens"""
    rigid = RigidDof(target.dofs)
    rigid.setPosition(start_position - rigid.forward * step * factor) 
    print(f"Up at ({factor:0.1}),  position is: {rigid.getPosition()}")

def runAnimation(target, step): 
    def startAnimation2(target, step, factor):
          # the start_position parameter needs a value, let's simply take it from the "current" position  
          animate(animation_up, {"target": target, 
          			 "start_position": RigidDof(target.dofs).getPosition(),  
          			 "step": step}, duration=0.05)
      
    animate(animation_left, {"target": target, "step": step}, duration=0.05, onDone=startAnimation2)
   
def createScene(root):
    root.addObject(AnimationManager(root))    
    root.addChild("actuators")
    
    root.actuators.addObject("MechanicalObject", template="Rigid3", name="dofs")
    runAnimation(root.actuators, 4.0)    
```
Which should print position ranging from [0,0,0] to [4,0,0] then from [4,0,0] to [4, 0, 4]

NB: there is also an additional problem in your code in the following lines :
```python
    animate(animation_left,{"actuators": actuators, "step": step}, duration=0.2, onDone=startAnimation2()
    animate(animation_left,{"actuators": actuators, "step": step}, duration=0.2, onDone=startAnimation2(factor=1.0, ...))
```    
because you wrote startAnimation() with the parenthesis. This is not supposed to work that way. The onDone of "animate" is taking a callable object as object, ... which means function itself and not the return value of its execution. So you should simply
```python
    animate(animation_left,{"actuators": actuators, "step": step}, duration=0.2, onDone=startAnimation2 ) # no parenthesis after startAnimation2
```

**Reply by dusevitch** [reply]
Thanks @damienmarchal, this is a great solution.
